store/sqlstore/thread_store.go

```diff
@@ -52,247 +52,278 @@ func (s *SqlThreadStore) Get(id string) (*model.Thread, error) {
 	return &thread, nil
 }
 
-func (s *SqlThreadStore) GetThreadsForUser(userId, teamId string, opts model.GetUserThreadsOpts) (*model.Threads, error) {
-	type JoinedThread struct {
-		PostId         string
-		ReplyCount     int64
-		LastReplyAt    int64
-		LastViewedAt   int64
-		UnreadReplies  int64
-		UnreadMentions int64
-		Participants   model.StringArray
-		model.Post
+// GetTotalUnreadThreads counts the number of unread threads for the given user, optionally
+// constrained to the given team + DMs/GMs.
+func (s *SqlThreadStore) GetTotalUnreadThreads(userId, teamId string, opts model.GetUserThreadsOpts) (int64, error) {
+	var totalUnreadThreads int64
+
+	query := s.getQueryBuilder().
+		Select("COUNT(DISTINCT(Posts.RootId))").
+		From("Posts").
+		LeftJoin("ThreadMemberships ON Posts.RootId = ThreadMemberships.PostId").
+		Where("Posts.CreateAt > ThreadMemberships.LastViewed").
+		Where(sq.Eq{
+			"ThreadMemberships.UserId":    userId,
+			"ThreadMemberships.Following": true,
+		})
+
+	if teamId != "" {
+		query = query.
+			LeftJoin("Channels ON Posts.ChannelId = Channels.Id").
+			Where(sq.Or{
+				sq.Eq{"Channels.TeamId": teamId},
+				sq.Eq{"Channels.TeamId": ""},
+			})
 	}
 
-	fetchConditions := sq.And{
-		sq.Eq{"ThreadMemberships.UserId": userId},
-		sq.Eq{"ThreadMemberships.Following": true},
+	if !opts.Deleted {
+		query = query.Where(sq.Eq{"COALESCE(Posts.DeleteAt, 0)": 0})
+	}
+
+	sql, args, err := query.ToSql()
+	if err != nil {
+		return 0, errors.Wrapf(err, "failed to build query to count unread threads for user id=%s", userId)
 	}
 
+	err = s.GetMasterX().Get(&totalUnreadThreads, sql, args...)
```

Comment:
Should we use the replica here? 

Same for other methods as well.

Reply:
Ooh, good catch! I don't see any reason we can't start using replicas here.

Reply:
However, I wonder if we might make this change holistically in a separate PR? I'd like to be able to isolate unexpected changes of behaviour, and it looks like much of `ThreadStore` will need to be touched.

Reply:
>However, I wonder if we might make this change holistically in a separate PR? I'd like to be able to isolate unexpected changes of behaviour, and it looks like much of ThreadStore will need to be touched.


SGTM